Fix issue with enum discriminators

issy opened the pull request:
When generating code from a schema like this
```
"Pet": {
  "type": "object",
  "properties": {
    "petType": {
      "type": "string"
    }
  },
  "required": ["petType"],
  "discriminator": {
    "propertyName": "petType",
    "mapping": {
      "dog": "#/components/schemas/Dog",
      "poodle": "#/components/schemas/Dog",
      "cat": "#/components/schemas/Cat"
    }
  }
}
```
I would expect to see the following TypeScript output
```ts
export type PetBase = { petType: string };
export type Dog = { petType: "dog" | "poodle" } & PetBase & { bark?: string };
export type Cat = { petType: "cat" } & PetBase & { name?: string };
export type Pet = Dog | Cat;
```
However, `oazapfts` generates the following instead
```ts
export type PetBase = { petType: string };
export type Dog = { petType: "dog" } & PetBase & { bark?: string };
export type Cat = { petType: "Cat" } & PetBase & { name?: string };
export type Lizard = { petType: "Lizard" } & PetBase & { lovesRocks?: boolean };
export type Pet = Dog | Cat | Lizard;
```
There is a problem with this - `Dog` can't be picked by the value `'poodle'`. It picks the "first" mapping entry instead of using all mapping entries. This has caused issues in my workflow, as the generated code for our web app ends up having incomplete type information. This issue has been logged in https://github.com/oazapfts/oazapfts/issues/674 and https://github.com/oazapfts/oazapfts/issues/611

In my case, my schema has 3 mapping entries pointing to the same schema. I'm sure you can see how this would cause issues.

Anyway, this PR aims to resolve that.

issy commented on packages/codegen/src/index.test.ts:
```diff
@@ -60,23 +60,28 @@ describe("generateSource", () => {
   });
 
-  it("should handle properties both inside and outside of allOf", async () => {
+  it(async () => {
```
Just nested these to deduplicate setup.

Xiphe replied:
Is it a common patter to nest `it` calls? Would have expected to use `describe` + `beforeEach|beforeAll` here.

issy commented on packages/codegen/src/index.test.ts:
```diff
@@ -60,23 +60,28 @@ describe("generateSource", () => {
+        'export type Cat = { petType: "cat"; } & PetBase & { name?: string; };',
+      );
+      expect(src).toContain(
+        'export type Dog = { petType: "dog" | "poodle"; } & PetBase & { bark?: string; };',
```
When there are more than 1 mapping matches, we get a string union

issy commented on packages/codegen/src/index.test.ts:
```diff
@@ -60,23 +60,28 @@ describe("generateSource", () => {
+      expect(src).toContain(
+        'export type Lizard = { petType: "Lizard"; } & PetBase & { lovesRocks?: boolean; };',
+      );
+      expect(src).toContain("export type Pet = Dog | Cat | Lizard;");
```
Deduplicated members here now.

Xiphe replied:
I read this as "Prior to the fix, there were duplication here". But we had the same assertion in line 73 before. 

So how is this test validating the fix, or something else?

Xiphe replied:
...assuming that the duplication was added along with the fix, right? So Current version doesn't have duplication but with the change from `find` to `filter` it was added, then mitigated using `_.uniq` so that this assertion remains valid, right?

Xiphe commented on packages/codegen/src/index.test.ts:
```diff
@@ -11,7 +11,7 @@ const demoFolder = path.join(rootFolder, "demo");
  * Generate an API from a relative path and convert it into a single line.
  */
-async function generate(file: string, opts: Opts = {}) {
+async function generate(file: string, opts: Opts = {}): Promise<string> {
```
nit: Let's use implicit return type

Xiphe reviewed (commented):
Hey @issy 

Thank you so much for taking the time to dig into this, really appreciated!

I have two nitty comments, you're invited to fix those or I can also quickly add them with the merge. 

I don't really see how this fixes https://github.com/oazapfts/oazapfts/issues/674, thought that issue was more about using the actual Enum type instead of literals when discriminating. When this is actually covered by this change I'd love to have a test-case for it.

 Also I don't see how this is fixing https://github.com/oazapfts/oazapfts/issues/611 as the `type` (or `petType` in this example) is still present - which is expected in this example using explicit discriminators but in the examples of the issue I read it as it's not expected as there is no explicit discriminator. Same here, when this change fixes the issue, I would love to see a test-case to proove it.
 
 In any case, this fixes the issue you're describing in the PR - happy to bring this in soon.

issy commented:
> I don't really see how this fixes #674, thought that issue was more about using the actual Enum type instead of literals when discriminating. When this is actually covered by this change I'd love to have a test-case for it.

@Xiphe correct, sort of. #674 is mostly about the actual Enum type (which this PR does not fix), but _part of_ that issue is the actual contents of the union, which this PR _does_ fix.

> Also I don't see how this is fixing #611 as the `type` (or `petType` in this example) is still present - which is expected in this example using explicit discriminators but in the examples of the issue I read it as it's not expected as there is no explicit discriminator. Same here, when this change fixes the issue, I would love to see a test-case to proove it.

Apologies for poorly referencing this issue. It seems I was actually referring more explicitly to issues brought up in the _comments_ of #611. The initial description in #611 doesn't elaborate on whether or not a discriminator is used, so I don't know if this will actually resolve that issue (I suspect a discriminator _is_ used, just going by the provided code output).

Furthermore, I'm not entirely sure if `Lizard` should still be included in the union, as it is not explicitly mentioned as a discriminator. I've kept it for now, as I suspect there are implementations in the wild that rely on this behaviour. I'm no OpenAPI expert, so maybe there is documentation explicitly stating that this discriminator should be implicitly included, as it is now.

I believe I've addressed your PR comments. Let me know if there's anything else!

Xiphe reviewed (approved):
LGTM :turtle:
